### PR TITLE
Avoid O(n) character accesses in String.UTF8View._foreignCount

### PR DESCRIPTION
**Pull request by Catfish-Man**
Fixes rdar://problem/50965043

### CODE REVIEW

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
Locally I see failures in 

UTF8ForContiguousUTF16/StringsWithUnpairedSurrogates
UTF8ForNonContiguousUTF16/utf16Tests

which probably implies that CFString handles surrogates slightly differently somehow.

**Comment by Catfish-Man** [comment]
@swift-ci please test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

**Comment by Catfish-Man** [comment]
@swift-ci please smoke benchmark

File: stdlib/public/core/StringUTF8View.swift [hunk 1]
```diff
@@ -470,14 +470,28 @@ extension String.UTF8View {
...
   internal func _foreignDistance(from i: Index, to j: Index) -> Int {
     _internalInvariant(_guts.isForeign)
+    
+    #if _runtime(_ObjC)
+    // Currently, foreign  means NSString
+    if let count = _cocoaStringUTF8Count(
+      _guts._object.cocoaObject,
+      range: i._encodedOffset ..< j._encodedOffset
+    ) {
+      let refinedCount = count - (i.transcodedOffset + j.transcodedOffset)
```
[the remark this answers is not in the document]
Comment (Catfish-Man) [reply stdlib/public/core/StringUTF8View.swift]: Yup, will do

File: stdlib/public/core/StringBridge.swift [hunk 2]
```diff
@@ -90,6 +90,26 @@ internal func _cocoaStringCopyUTF8(
...
+    _swift_shims_CFRange(location: range.startIndex, length: range.count),
+    kCFStringEncodingUTF8,
+    0,
+    0,
+    UnsafeMutablePointer<UInt8>(Builtin.inttoptr_Word(0._builtinWordValue)),
+    0,
+    &count
+  )
+  return converted == len ? count : nil
+}
```
[the remark this answers is not in the document]
Comment (Catfish-Man) [reply stdlib/public/core/StringBridge.swift]: Random NSString subclasses are free to do as they please, including just being incorrect. CFString itself is either ASCII or UTF16 though. Not 100% certain if that answers the question?

**Comment by Catfish-Man** [comment]
@swift-ci please smoke test

File: stdlib/public/core/StringUTF8View.swift (hunk 1 above)
[the remark this answers is not in the document]
Comment (milseman) [reply stdlib/public/core/StringUTF8View.swift]: Spaces after `//`

**Pull request merged**